### PR TITLE
refactor(metadata) : operations with new dir entry design

### PR DESCRIPTION
**Pull request by xiaguan**
In this pull request ,I removed the old direntry.. And refactor the metadta's operations with new designed dir entry.
And I fixed a bug caused by setattr().

### CODE REVIEW

File: src/async_fuse/memfs/kv_engine/etcd_impl.rs [hunk 1]
```diff
@@ -271,6 +271,18 @@ impl MetaTxn for EtcdTxn {
...
+            let prev_value = self.buffer.get(&key);
+            if let Some(prev_value) = prev_value {
+                // If the value is None, it means that the key is deleted in the same
+                // transaction.
+                assert!(
+                    prev_value.is_none(),
+                    "set the key={key:?} after delete in the same transaction"
+                );
+            }
+        }
```
[the remark this answers is not in the document]
Comment (xiaguan) [reply src/async_fuse/memfs/kv_engine/etcd_impl.rs:285]: It means you set a key after you delete it in ssame transaction.It's strange. I think overwrite same key twice is a logical bug,too. I will fix it.

File: src/async_fuse/memfs/s3_metadata.rs [hunk 2]
```diff
@@ -690,33 +699,29 @@ impl<S: S3BackEnd + Sync + Send + 'static> MetaData for S3MetaData<S> {
     ) -> DatenLordResult<(Duration, FuseAttr, u64)> {
         retry_txn!(TXN_RETRY_LIMIT, {
             let mut txn = self.kv_engine.new_meta_txn().await;
-            let parent_node = self.get_inode_from_txn(txn.as_mut(), parent).await?;
-            let Some(child_entry) = parent_node.get_entry(child_name) else {
+            if NEED_CHECK_PERM {
+                let parent_node = self.get_inode_from_txn(txn.as_mut(), parent).await?;
+                parent_node
+                    .get_attr()
+                    .check_perm(context.user_id, context.group_id, 1)?;
```
[the remark this answers is not in the document]
Comment (xiaguan) [reply src/async_fuse/memfs/s3_metadata.rs:706]: The read permission of a dir is read its all dir entries. Lookup doesn't need it.

File: src/async_fuse/memfs/kv_engine/etcd_impl.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Reto911) [reply src/async_fuse/memfs/kv_engine/etcd_impl.rs:285]: So both we set a key after deleting it and set it twice are error, right?

However, I think you already do so in <https://github.com/datenlord/datenlord/blob/080834dc08c57be9869283ac6865e7d9cdf64f05/src/async_fuse/memfs/s3_metadata.rs#L827-L847>.

[the remark this answers is not in the document]
Comment (xiaguan) [reply src/async_fuse/memfs/kv_engine/etcd_impl.rs]: Fixed.

File: src/async_fuse/memfs/s3_metadata.rs [hunk 3]
```diff
@@ -748,12 +754,12 @@ impl<S: S3BackEnd + Sync + Send + 'static> MetaData for S3MetaData<S> {
...
 
-        let build_enoent = |name: &str, parent: INum, parent_name: &str| {
+        let build_enoent = |name: &str, parent: INum, child_name: &str| {
             build_error_result_from_errno(
                 Errno::ENOENT,
                 format!(
                     "exchange_pre_check() failed to find child entry of name={name:?} \
-                        under parent directory ino={parent} and name={parent_name:?}"
+                        under parent directory ino={parent} and name={child_name:?}"
                 ),
```
[the remark this answers is not in the document]
Comment (xiaguan) [reply src/async_fuse/memfs/s3_metadata.rs]: Fixed

File: src/async_fuse/memfs/s3_node.rs [hunk 4]
```diff
@@ -144,29 +135,9 @@ impl<S: S3BackEnd + Send + Sync + 'static> S3Node<S> {
         meta: &S3MetaData<S>,
     ) -> BoxFuture<'_, DatenLordResult<S3Node<S>>> {
```
[the remark this answers is not in the document]
Comment (xiaguan) [reply src/async_fuse/memfs/s3_node.rs]: Fixed related issues.

File: src/async_fuse/memfs/s3_node.rs [hunk 5]
```diff
@@ -241,6 +212,19 @@ impl<S: S3BackEnd + Send + Sync + 'static> S3Node<S> {
...
+        match self.data {
+            S3NodeData::Directory => debug_assert_eq!(new_attr.kind, SFlag::S_IFDIR),
+            S3NodeData::RegFile(..) => debug_assert_eq!(new_attr.kind, SFlag::S_IFREG),
+            S3NodeData::SymLink(..) => debug_assert_eq!(new_attr.kind, SFlag::S_IFLNK),
+        }
+
+        self.attr.write().clone_from(&new_attr);
+        old_attr
+    }
+
```
[the remark this answers is not in the document]
Comment (xiaguan) [reply src/async_fuse/memfs/s3_node.rs]: Removed

**Review by Reto911 — DISMISSED** [review]
LGTM

File: src/async_fuse/memfs/s3_metadata.rs [hunk 6]
```diff
@@ -725,14 +730,15 @@ impl<S: S3BackEnd + Sync + Send + 'static> MetaData for S3MetaData<S> {
...
-    /// Rename helper for exchange rename
+    #[allow(clippy::too_many_lines)] // TODO: refactor it into smaller functions
     #[instrument(skip(self), err, ret)]
     async fn rename(&self, context: ReqContext, param: RenameParam) -> DatenLordResult<()> {
         let old_parent = param.old_parent;
         let old_name = param.old_name.as_str();
         let new_parent = param.new_parent;
         let new_name = param.new_name.as_str();
         let flags = param.flags;
+        // TODO: replace the new_node should delete its related data
```
Comment (Reto911) [inline src/async_fuse/memfs/s3_metadata.rs:741]: Are you going to fix delete related issue in this PR?

Comment (Reto911) [reply src/async_fuse/memfs/s3_metadata.rs:741]: #456, resolved in #463, but overwritten in this PR.

File: src/async_fuse/memfs/mod.rs
Comment (Reto911) [inline src/async_fuse/memfs/mod.rs:1]: What is the `dead_code` in mod `direntry`?

Comment (xiaguan) [reply src/async_fuse/memfs/mod.rs:1]: fixed

**Pull request merged**